File: codes/Misc/OEIS.cpp
// 若一個線性遞迴有 k 項，給他恰好 2*k 個項可以求出線性遞迴
// f915c2
template <typename T>
vector<T> BerlekampMassey(vector<T> a) {
    auto scalarProduct = [](vector<T> v, T c) {
        for (T &x: v) x *= c;
        return v;
    };
    vector<T> s, best;
    int bestPos = 0;
    for (size_t i = 0; i < a.size(); i++) {
        T error = a[i];
        for (size_t j = 0; j < s.size(); j++) error -= s[j] * a[i-1-j];
        if (error == 0) continue;
        if (s.empty()) {
            s.resize(i + 1);
            bestPos = i;
            best.push_back(1 / error);
            continue;
        }
        vector<T> fix = scalarProduct(best, error);
        fix.insert(fix.begin(), i - bestPos - 1, 0);
        if (fix.size() >= s.size()) {
            best = scalarProduct(s, - 1 / error);
            best.insert(best.begin(), 1 / error);
            bestPos = i;
            s.resize(fix.size());
        }
        for (size_t j = 0; j < fix.size(); j++)
            s[j] += fix[j];
    }
    return s;
}

File: codes/Misc/OEIS.py
from fractions import Fraction

def BerlekampMassey(a: list[Fraction]) -> list[Fraction]:
    def scale(v: list[Fraction], c: Fraction) -> list[Fraction]:
        return [x * c for x in v]

    s: list[Fraction] = []
    best: list[Fraction] = []
    bestPos = 0

    for i in range(len(a)):
        error: Fraction = a[i]
        for j in range(len(s)):
            error -= s[j] * a[i - 1 - j]
        if error == 0:
            continue

        if not s:
            s = [Fraction(0)] * (i + 1)
            bestPos = i
            best = [Fraction(1, error)]
            continue

        fix = scale(best, error)
        fix = [Fraction(0)] * (i - bestPos - 1) + fix

        if len(fix) >= len(s):
            best = scale(s, Fraction(-1, error))
            best.insert(0, Fraction(1, error))
            bestPos = i
            if len(s) < len(fix):
                s += [Fraction(0)] * (len(fix) - len(s))

        for j in range(len(fix)):
            s[j] += fix[j]

    return list(reversed(s))

n = int(input())
l = list(map(Fraction, input().split()))
for i in range(len(l)):
    coeffs = BerlekampMassey(l[:i+1])
    for x in coeffs:
        print(x, end=" ")
    print()